app/apollo/resolvers/common.js

```diff
@@ -185,6 +225,36 @@ const validAuth = async (me, org_id, action, type, queryName, context, attrs = n
   }
 };
 
+/*
+Check if cache exists and is authorized for each resource of passed 'type'
+Return true if user is authorized for every resource of passed 'type' or if cache doesn't exist
+Return false if user is not authorized for every resource of passed 'type'
+*/
+const cacheAllAllowed = async (me, org_id, action, type, queryName, context) => {
+  const {models} = context;
+
+  if (context.recoveryHintsMap) {
+    context['recoveryHints'] = context.recoveryHintsMap[queryName];
+  }
+
+  // razeedash users (x-api-key)
+  if(me && me.type == 'userToken'){
+    const result = await models.User.userTokenIsAuthorized(me, org_id, action, type, context);
+    if(!result){
+      throw new RazeeForbiddenError(
+        context.req.t('You are not allowed to {{action}} on {{type}} under organization {{org_id}} for the query {{queryName}}.', {'action':action, 'type':type, 'org_id':org_id, 'queryName':queryName, interpolation: { escapeValue: false }}
+        ), context);
+    }
+    return true;
+  }
+
+  //when IAM permissions are given to read Channel and a specific named channel this error should not happen
+  if (me === null || !(await models.User.isAuthorized(me, org_id, action, type, null, context, true))) {
+    return false;
+  }
+  return true;
+};
+
```

Comment:
Function impl incorrect/problematic/incomplete:
- `recoveryHints` code purpose is not clear -- should explain why it's needed if present.
- Sometimes will call `userTokenIsAuthorized` which will do more than checking cache
- Doesn't just return true/false, sometimes explicitly throws exception
- **NOTE**: As noted in at least one other review comment (cluster.js), the design of this function appears to be incorrect however -- as described, it's impossible to tell the difference between "cache says user **is** authorized on all" and "cache doesn't know", yet is used to determine whether fine-grained-auth checks can later be skipped.  Fine grained auth checks can only be skipped if user **is** authorized on all, not on cache miss.

Reply:
recoveryHints is not finished... What it would do if finished would give a forbiddenError and say "ForbiddenError: 'You do not have enough permission to list all the Kubernetes resources in the account. Scope your request with the cluster or subscription for resources that you do have permissions to list, and try again." This seems useless with the new usage of validAuth and cacheAllAllowed. These two functions now separate between being specific with attributes (validAuth) and being generalized to a resource type (cacheAllAllowed). I will remove this.

Reply:
In a similar vein, userTokenIsAuthorized is not used in the case of cacheAllAllowed but could be used in the specific case in validAuth() if doing userToken tests.